mecanum_drive_controller/src/mecanum_drive_controller.cpp

```diff
@@ -0,0 +1,502 @@
+  }
+  controller_state_publisher_->unlock();
+
+  RCLCPP_INFO(get_node()->get_logger(), "configure successful");
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+void MecanumDriveController::reference_callback(const std::shared_ptr<ControllerReferenceMsg> msg)
+{
+  // if no timestamp provided use current time for command timestamp
+  if (msg->header.stamp.sec == 0 && msg->header.stamp.nanosec == 0u)
+  {
+    RCLCPP_WARN(
+      get_node()->get_logger(),
+      "Timestamp in header is missing, using current time as command "
+      "timestamp.");
+    msg->header.stamp = get_node()->now();
+  }
+  const auto age_of_last_command = get_node()->now() - msg->header.stamp;
+
+  if (ref_timeout_ == rclcpp::Duration::from_seconds(0) || age_of_last_command <= ref_timeout_)
+  {
+    input_ref_.writeFromNonRT(msg);
+  }
+  else
+  {
+    RCLCPP_ERROR(
+      get_node()->get_logger(),
+      "Received message has timestamp %.10f older for %.10f which is more then allowed timeout "
+      "(%.4f).",
+      rclcpp::Time(msg->header.stamp).seconds(), age_of_last_command.seconds(),
+      ref_timeout_.seconds());
+    reset_controller_reference_msg(msg, get_node());
+  }
+}
+
+controller_interface::InterfaceConfiguration
+MecanumDriveController::command_interface_configuration() const
+{
+  controller_interface::InterfaceConfiguration command_interfaces_config;
+  command_interfaces_config.type = controller_interface::interface_configuration_type::INDIVIDUAL;
+
+  command_interfaces_config.names.reserve(params_.command_joint_names.size());
+  for (const auto & joint : params_.command_joint_names)
+  {
+    command_interfaces_config.names.push_back(joint + "/" + params_.interface_name);
+  }
+
+  return command_interfaces_config;
+}
+
+controller_interface::InterfaceConfiguration MecanumDriveController::state_interface_configuration()
+  const
+{
+  controller_interface::InterfaceConfiguration state_interfaces_config;
+  state_interfaces_config.type = controller_interface::interface_configuration_type::INDIVIDUAL;
+
+  state_interfaces_config.names.reserve(state_joint_names_.size());
+
+  for (const auto & joint : state_joint_names_)
+  {
+    state_interfaces_config.names.push_back(joint + "/" + params_.interface_name);
+  }
+
+  return state_interfaces_config;
+}
+
+std::vector<hardware_interface::CommandInterface>
+MecanumDriveController::on_export_reference_interfaces()
+{
+  reference_interfaces_.resize(NR_REF_ITFS, std::numeric_limits<double>::quiet_NaN());
+
+  std::vector<hardware_interface::CommandInterface> reference_interfaces;
+
+  reference_interfaces.reserve(reference_interfaces_.size());
+
+  for (size_t i = 0; i < reference_interfaces_.size(); ++i)
+  {
+    reference_interfaces.push_back(hardware_interface::CommandInterface(
+      get_node()->get_name(), reference_names_[i] + "/" + params_.interface_name,
+      &reference_interfaces_[i]));
+  }
```

Comment:
This is how I would propose to handle this – when the parameter is deleted. Are all tree interfaces velocity here?

```suggestion
  std::vector<hardware_interface::CommandInterface> reference_interfaces;
  reference_interfaces.reserve(reference_interfaces_.size());
  
  std::vector<std::string> reference_interface_names = {"linear/x/velocity", "linear/y/velocity", "angular/z/velocity"}

  for (size_t i = 0; i < reference_interfaces_.size(); ++i)
  {
    reference_interfaces.push_back(hardware_interface::CommandInterface(
      get_node()->get_name(), reference_interface_names[i],
      &reference_interfaces_[i]));
  }
```

Reply:
then we also dont need to set the parameter interface_name_ ?? or just leave it there

i wrote it like this at the moment
  reference_interfaces.push_back(hardware_interface::CommandInterface(
    get_node()->get_name(), std::string("linear_x/") + params_.interface_name,
    &reference_interfaces_[0]));
  reference_interfaces.push_back(hardware_interface::CommandInterface(
    get_node()->get_name(), std::string("linear_y/") + params_.interface_name,
    &reference_interfaces_[1]));
  reference_interfaces.push_back(hardware_interface::CommandInterface(
    get_node()->get_name(), std::string("angular_z/") + params_.interface_name,
    &reference_interfaces_[2]));